### PR TITLE
dbengine extent cache

### PR DESCRIPTION
**Pull request by mfundul**
##### Component Name
database/engine
##### Test Plan

Choose a big enough time period in the netdata dashboard and create a snapshot.

Measure disk I/O and time to completion. Compare with the code in master.

You can measure disk I/O reads before and after the experiment by looking at `/proc/diskstats` and subtracting the values of the 6th column. Those are 512-byte sectors.

##### Additional Information
- The typical netdata query will only fetch 1 dimension from disk for a specific chart.
- Subsequent queries will fetch the rest of the dimensions.
- As a result, the extents will be read from disk multiple times.
A small extent cache will speed this up.

### CODE REVIEW

File: database/engine/rrdengine.c [hunk 1]
```diff
@@ -99,6 +273,34 @@ void read_extent_cb(uv_fs_t* req)
...
         /* care, we don't hold the descriptor mutex */
     }
+    {
+        uint8_t xt_is_cached = 0, xt_is_inflight = 0;
+        unsigned xt_idx;
+        struct extent_info *extent = xt_io_descr->descr_array[0]->extent;
+
+        xt_is_cached = !lookup_in_xt_cache(wc, extent, &xt_idx);
+        xt_is_inflight = check_bit(wc->xt_cache.inflight_bitmap, xt_idx);
+        if (xt_is_cached && xt_is_inflight) {
```
Comment (stelfrag) [inline database/engine/rrdengine.c:283]: xt_idx is undefined if xt_is_cache is false

Comment (mfundul) [reply database/engine/rrdengine.c]: It should be safe in practice but it's quite ugly so I'm fixing it.

**Review by stelfrag — APPROVED** [review]

File: database/engine/rrdengine.h [hunk 2]
```diff
@@ -99,6 +100,27 @@ struct generic_io_descriptor {
     struct completion *completion;
 };
 
+struct extent_cache_element {
+    struct extent_info *extent; /* The ABA problem is avoided with the help of fileno below */
+    unsigned fileno;
+    struct extent_cache_element *prev; /* LRU */
+    struct extent_cache_element *next; /* LRU */
```
Comment (underhood) [inline database/engine/rrdengine.h:107]: /* MRU */?

Comment (mfundul) [reply database/engine/rrdengine.h:107]: Hmm, maybe I should just say `LRU list`?

**Review by underhood — APPROVED** [review]
all comments i would have are of "subjective" nature, so approving :)

**Pull request merged**